### PR TITLE
[Feature] add Tensor Parallelism to Omnigen2

### PR DESCRIPTION
**Pull request by zzhuoxin1508**
PLEASE FILL IN THE PR DESCRIPTION HERE ENSURING ALL CHECKLIST ITEMS (AT THE BOTTOM) HAVE BEEN CONSIDERED.

## Purpose
Add Tensor Parallelism to  omnigen2.
## Test Plan
python image_edit.py \
    --image /workspace/image1.png \
    --model "OmniGen2/OmniGen2" \
    --prompt "Change the background to classroom." \
    --negative-prompt "(((deformed))), blurry, over saturation, bad anatomy, disfigured, poorly drawn face, mutation, mutated, (extra_limb), (ugly), (poorly drawn hands), fused fingers, messy drawing, broken legs censor, censored, censor_bar" \
    --num-inference-steps 50 \
    --seed 0 \
    --guidance-scale 5.0 \
    --guidance-scale-2 2.0 \
    --output /workspace/outputs/image_edit.png

## Test Result
| TP Size | Generation Time | Reduction (Time) | Peak VRAM | Generated Image |
|---------|----------------|-----------------|-----------|-----------------|
| TP=1    | 70.15s     | -               | ~20.9 GiB | image_editp1t  |
| TP=2    | 55.2s     | ↓ 21.3%         | ~17.9 GiB |image_edittp2   |

## Note
Add Tensor Parallelism to  omnigen2.The attention layers are not tensor-parallelized , This is because the
 model's actual Q/KV head counts (21 and 7) . Only the MLP layers (`gate_up_proj`, `down_proj`) use tensor parallelism.
 
---

<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan. Please provide the test scripts & test commands. Please state the reasons if your codes don't require additional test scripts. For test file guidelines, please check the [test style doc](https://docs.vllm.ai/projects/vllm-omni/en/latest/contributing/ci/tests_style/)
- [ ] The test results. Please paste the results comparison before and after, or the e2e results.
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model. **Please run `mkdocs serve` to sync the documentation editions to `./docs`.**
- [ ] (Optional) Release notes update. If your change is user-facing, please update the release notes draft.
</details>

**BEFORE SUBMITTING, PLEASE READ <https://github.com/vllm-project/vllm-omni/blob/main/CONTRIBUTING.md>** (anything written below this line will be removed by GitHub Actions)

### CODE REVIEW

**Comment by zzhuoxin1508** [comment]
original image 
Gemini_Generated_Image_38otyo38otyo38ot

**Comment by zzhuoxin1508** [comment]
@princepride

File: vllm_omni/diffusion/models/omnigen2/omnigen2_transformer.py [hunk 1]
```diff
@@ -21,8 +23,10 @@
 logger = logging.getLogger(__name__)
 
 
-def swiglu(x, y):
-    return F.silu(x.float(), inplace=False).to(x.dtype) * y
+class SwiGLU(nn.Module):
```
Comment (princepride) [inline vllm_omni/diffusion/models/omnigen2/omnigen2_transformer.py]: Can we use action function from `from vllm.model_executor.layers.activation import get_act_and_mul_fn, get_act_fn`

File: vllm_omni/diffusion/models/omnigen2/omnigen2_transformer.py [hunk 2]
```diff
@@ -78,22 +81,19 @@ def forward(
...
+        kv_size = self.num_kv_heads * self.head_dim
+        query, key, value = qkv.split([q_size, kv_size, kv_size], dim=-1)
 
         # Reshape tensors for attention computation
-        query = query.view(batch_size, sequence_length, self.num_heads, self.head_dim)
-        key = key.view(batch_size, sequence_length, self.num_kv_heads, self.head_dim)
-        value = value.view(batch_size, sequence_length, self.num_kv_heads, self.head_dim)
+        query = query.unflatten(-1, (self.num_heads, -1))
+        key = key.unflatten(-1, (self.num_kv_heads, -1))
+        value = value.unflatten(-1, (self.num_kv_heads, -1))
```
Comment (princepride) [inline vllm_omni/diffusion/models/omnigen2/omnigen2_transformer.py:90]: This code seems to make matrix dimension changes safer, but I have a question: under what circumstances would the original code fail, I believe our matrix's dimension shouldn't change?

File: vllm_omni/diffusion/models/omnigen2/omnigen2_transformer.py
Comment (princepride) [inline vllm_omni/diffusion/models/omnigen2/omnigen2_transformer.py]: Can you help me remove all `initialize_weights`, these code only used in training, I believe previous code didn't be fully reviewed before merging.

File: vllm_omni/diffusion/models/omnigen2/omnigen2_transformer.py (hunk 2 above)
Comment (zzhuoxin1508) [reply vllm_omni/diffusion/models/omnigen2/omnigen2_transformer.py:90]: our matrix dimensions won't change here. The main benefit of unflatten is just cleaner code — no need to explicitly pass batch_size and sequence_length.

File: vllm_omni/diffusion/models/omnigen2/omnigen2_transformer.py
Comment (zzhuoxin1508) [reply vllm_omni/diffusion/models/omnigen2/omnigen2_transformer.py]: done

File: vllm_omni/diffusion/models/omnigen2/omnigen2_transformer.py (hunk 1 above)
Comment (zzhuoxin1508) [reply vllm_omni/diffusion/models/omnigen2/omnigen2_transformer.py]: done

File: vllm_omni/diffusion/models/omnigen2/omnigen2_transformer.py [hunk 3]
```diff
@@ -5,14 +5,16 @@
...
 import torch
 import torch.nn as nn
-import torch.nn.functional as F
 from diffusers.models.activations import get_activation
 from diffusers.models.embeddings import Timesteps, get_1d_rotary_pos_embed
 from diffusers.models.modeling_outputs import Transformer2DModelOutput
 from einops import rearrange, repeat
-from torch.nn import RMSNorm
+from vllm.model_executor.layers.activation import get_act_and_mul_fn
+from vllm.model_executor.layers.layernorm import RMSNorm
```
Comment (lishunyang12) [inline vllm_omni/diffusion/models/omnigen2/omnigen2_transformer.py:13]: Switching from `torch.nn.RMSNorm` to vllm's `RMSNorm` breaks `LuminaLayerNormContinuous` — it passes `elementwise_affine=` on the `norm_type="rms_norm"` branch (around line 296), but vllm's `RMSNorm` doesn't accept that kwarg (it uses `has_weight` instead). That path isn't hit with the current OmniGen2 config, but it'll crash if anyone ever sets `norm_type="rms_norm"`. Either keep `torch.nn.RMSNorm` for the non-TP norms, or fix that call site to use `has_weight=elementwise_affine`.

File: vllm_omni/diffusion/models/omnigen2/omnigen2_transformer.py [hunk 4]
```diff
@@ -340,32 +326,31 @@ def __init__(
...
         super().__init__()
-        self.swiglu = swiglu
 
         # custom hidden_size factor multiplier
         if ffn_dim_multiplier is not None:
             inner_dim = int(ffn_dim_multiplier * inner_dim)
         inner_dim = multiple_of * ((inner_dim + multiple_of - 1) // multiple_of)
 
-        self.linear_1 = nn.Linear(
+        self.gate_up_proj = MergedColumnParallelLinear(
```
Comment (lishunyang12) [inline vllm_omni/diffusion/models/omnigen2/omnigen2_transformer.py:335]: Neither `gate_up_proj` nor `down_proj` receives `quant_config` or `prefix`. Compare with the HunyuanImage3 MLP which threads both through. Not blocking for a TP-only PR, but quantization will silently skip these layers without it.

File: vllm_omni/diffusion/models/omnigen2/omnigen2_transformer.py [hunk 5]
```diff
@@ -340,32 +326,31 @@ def __init__(
...
             dim,
-            inner_dim,
+            [inner_dim, inner_dim],
             bias=False,
+            return_bias=False,
         )
-        self.linear_2 = nn.Linear(
+        self.act_fn = get_act_and_mul_fn("silu")
```
Comment (lishunyang12) [inline vllm_omni/diffusion/models/omnigen2/omnigen2_transformer.py:341]: Minor: the old `swiglu` did `F.silu(x.float()).to(x.dtype)` — explicit fp32 upcast for numerical stability. `SiluAndMul` operates in native dtype (or uses a fused CUDA kernel). Worth confirming this doesn't regress quality in bf16.

File: vllm_omni/diffusion/models/omnigen2/omnigen2_transformer.py [hunk 6]
```diff
@@ -1197,6 +1139,9 @@ def load_weights(self, weights: Iterable[tuple[str, torch.Tensor]]) -> set[str]:
                 weight_loader(param, loaded_weight, shard_id)
                 break
             else:
+                # feed_forward.linear_2 -> feed_forward.down_proj rename
+                if ".feed_forward.linear_2." in name:
```
Comment (lishunyang12) [inline vllm_omni/diffusion/models/omnigen2/omnigen2_transformer.py:1143]: Nit: this inline rename is easy to miss. Why not add it to a dedicated `rename_mapping` dict like other models do, instead of burying it in the else branch?

**Comment by princepride** [comment]
@lishunyang12 I have test on my local device, the results are same.

**Comment by princepride** [comment]
Config | Time | Speed Up | Memory Usage | Results
-- | -- | -- | -- | --
TP=1 | 19.82s | - | 14.59 GiB (single card) | image
TP=2 | 16.99s | ↓ 14.3% | 11.90 GiB (each card) | image
